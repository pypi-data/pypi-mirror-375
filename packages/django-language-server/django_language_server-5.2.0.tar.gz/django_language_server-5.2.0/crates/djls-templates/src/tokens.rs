use serde::Serialize;

#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum TokenType {
    Comment(String, String, Option<String>),
    DjangoBlock(String),
    DjangoVariable(String),
    Eof,
    HtmlTagOpen(String),
    HtmlTagClose(String),
    HtmlTagVoid(String),
    Newline,
    ScriptTagOpen(String),
    ScriptTagClose(String),
    StyleTagOpen(String),
    StyleTagClose(String),
    Text(String),
    Whitespace(usize),
}

impl TokenType {
    pub fn len(&self) -> usize {
        match self {
            TokenType::DjangoBlock(s)
            | TokenType::DjangoVariable(s)
            | TokenType::HtmlTagOpen(s)
            | TokenType::HtmlTagClose(s)
            | TokenType::HtmlTagVoid(s)
            | TokenType::ScriptTagOpen(s)
            | TokenType::ScriptTagClose(s)
            | TokenType::StyleTagOpen(s)
            | TokenType::StyleTagClose(s)
            | TokenType::Text(s) => s.len(),
            TokenType::Comment(content, _, _) => content.len(),
            TokenType::Whitespace(n) => *n,
            TokenType::Newline => 1,
            TokenType::Eof => 0,
        }
    }
}

#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Token {
    #[allow(clippy::struct_field_names)]
    token_type: TokenType,
    line: usize,
    start: Option<usize>,
}

impl Token {
    pub fn new(token_type: TokenType, line: usize, start: Option<usize>) -> Self {
        Self {
            token_type,
            line,
            start,
        }
    }

    pub fn lexeme(&self) -> String {
        match &self.token_type {
            TokenType::Comment(_, start, end) => match end {
                Some(end) => format!("{} {} {}", start, self.content(), end),
                None => format!("{} {}", start, self.content()),
            },
            TokenType::DjangoBlock(_) => format!("{{% {} %}}", self.content()),
            TokenType::DjangoVariable(_) => format!("{{{{ {} }}}}", self.content()),
            TokenType::Eof => String::new(),
            TokenType::HtmlTagOpen(_)
            | TokenType::ScriptTagOpen(_)
            | TokenType::StyleTagOpen(_) => format!("<{}>", self.content()),
            TokenType::HtmlTagClose(_)
            | TokenType::StyleTagClose(_)
            | TokenType::ScriptTagClose(_) => format!("</{}>", self.content()),
            TokenType::HtmlTagVoid(_) => format!("<{}/>", self.content()),
            TokenType::Newline | TokenType::Text(_) | TokenType::Whitespace(_) => self.content(),
        }
    }

    pub fn content(&self) -> String {
        match &self.token_type {
            TokenType::Comment(s, _, _)
            | TokenType::DjangoBlock(s)
            | TokenType::DjangoVariable(s)
            | TokenType::Text(s)
            | TokenType::HtmlTagOpen(s)
            | TokenType::HtmlTagClose(s)
            | TokenType::HtmlTagVoid(s)
            | TokenType::ScriptTagOpen(s)
            | TokenType::ScriptTagClose(s)
            | TokenType::StyleTagOpen(s)
            | TokenType::StyleTagClose(s) => s.to_string(),
            TokenType::Whitespace(len) => " ".repeat(*len),
            TokenType::Newline => "\n".to_string(),
            TokenType::Eof => String::new(),
        }
    }

    pub fn token_type(&self) -> &TokenType {
        &self.token_type
    }

    pub fn line(&self) -> &usize {
        &self.line
    }

    pub fn start(&self) -> Option<u32> {
        self.start
            .map(|s| u32::try_from(s).expect("Start position should fit in u32"))
    }

    pub fn length(&self) -> u32 {
        u32::try_from(self.token_type.len()).expect("Token length should fit in u32")
    }

    pub fn is_token_type(&self, token_type: &TokenType) -> bool {
        &self.token_type == token_type
    }
}

#[salsa::tracked]
pub struct TokenStream<'db> {
    #[tracked]
    #[returns(ref)]
    pub stream: Vec<Token>,
}

impl<'db> TokenStream<'db> {
    /// Check if the token stream is empty
    pub fn is_empty(self, db: &'db dyn crate::db::Db) -> bool {
        self.stream(db).is_empty()
    }

    /// Get the number of tokens
    pub fn len(self, db: &'db dyn crate::db::Db) -> usize {
        self.stream(db).len()
    }
}
