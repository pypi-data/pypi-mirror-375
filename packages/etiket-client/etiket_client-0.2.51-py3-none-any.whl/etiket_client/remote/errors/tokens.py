from etiket_client.remote.errors.meta import EtiketExceptionsMeta

class InvalidRequest(Exception):
    """The request is missing a required parameter, includes an invalid parameter value, or is otherwise malformed."""

class InvalidClient(Exception):
    """Client authentication failed."""

class InvalidGrant(Exception):
    """The provided authorization grant is invalid, expired, revoked, or was issued to another client."""

class UnauthorizedClient(Exception):
    """The authenticated client is not authorized to use this authorization grant type."""

class UnsupportedGrantType(Exception):
    """The authorization grant type is not supported by the authorization server."""

class InvalidScope(Exception):
    """The requested scope is invalid, unknown, malformed, or exceeds the scope granted."""

class AccessDenied(Exception):
    """The resource owner or authorization server denied the request."""

class UnsupportedResponseType(Exception):
    """The authorization server does not support obtaining an authorization code using this method."""

class TemporarilyUnavailable(Exception):
    """The authorization server is currently unable to handle the request."""

class TokenInvalid(Exception):
    """The provided token is invalid."""

class TokenRevoked(Exception):
    """The provided token has been revoked."""

class TokenUserCreateFailed(Exception):
    """Failed to create user from token information."""

class OAuth2Errors(metaclass=EtiketExceptionsMeta):
    invalid_request = InvalidRequest
    invalid_client = InvalidClient
    invalid_grant = InvalidGrant
    unauthorized_client = UnauthorizedClient
    unsupported_grant_type = UnsupportedGrantType
    invalid_scope = InvalidScope
    access_denied = AccessDenied
    unsupported_response_type = UnsupportedResponseType
    temporarily_unavailable = TemporarilyUnavailable

class TokenAuthErrors(metaclass=EtiketExceptionsMeta):
    token_invalid = TokenInvalid
    token_revoked = TokenRevoked
    token_user_create_failed = TokenUserCreateFailed