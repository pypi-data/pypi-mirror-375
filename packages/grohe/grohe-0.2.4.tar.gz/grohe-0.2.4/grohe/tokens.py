import httpx
import urllib.parse
from bs4 import BeautifulSoup

from grohe.dto.grohe_dto import GroheTokensDTO


class GroheTokens:
    def __init__(self, httpx_client: httpx.AsyncClient, api_url: str):
        self.__client = httpx_client
        self.__api_url = api_url


    async def get_tokens_from_credentials(self, grohe_email: str, grohe_password: str) -> GroheTokensDTO:
        """
        Get the initial access and refresh tokens from the given Grohe credentials.
        Args:
            grohe_email: The Grohe email.
            grohe_password: The Grohe password.

        Returns: A dict with the tokens.
        """
        response = await self.__client.get(f'{self.__api_url}/oidc/login', follow_redirects=True)
        response.raise_for_status()
        html = response.text

        soup = BeautifulSoup(html, 'html.parser')
        form = soup.find('form')
        if not form or 'action' not in form.attrs:
            raise Exception('Login form target URL not found')

        action_url = urllib.parse.urljoin(f'{self.__api_url}/oidc/login', form['action'])

        payload = {
            'username': grohe_email,
            'password': grohe_password,
        }

        headers = {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Referer': f'{self.__api_url}/oidc/login',
        }

        try:
            response = await self.__client.post(
                action_url, data=payload, headers=headers, follow_redirects=False
            )
            response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            if exc.response.status_code == 302:
                location = exc.response.headers.get('Location')
                if not location:
                    raise Exception('No redirect location found after login')

                tokens_url = location.replace('ondus://', 'https://')

                response = await self.__client.get(tokens_url)
                response.raise_for_status()
                json_data = response.json()

                tokens = GroheTokensDTO.from_dict(json_data)
                return tokens
            else:
                raise

        raise Exception(
            'Invalid username/password or unexpected response from Grohe service'
        )

    async def get_refresh_tokens(self, refresh_token: str) -> GroheTokensDTO:
        """
        Refresh the access and refresh tokens.
        Args:
            refresh_token: The refresh token.

        Returns: A dict with the new tokens.
        """
        data = {'refresh_token': refresh_token, 'grant_type': 'refresh_token'}
        response = await self.__client.post(f'{self.__api_url}/oidc/refresh', json=data)
        response.raise_for_status()
        json_data = response.json()

        tokens = GroheTokensDTO.from_dict(json_data)
        return tokens
