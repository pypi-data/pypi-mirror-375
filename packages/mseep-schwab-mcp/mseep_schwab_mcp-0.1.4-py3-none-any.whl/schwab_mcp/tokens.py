#

import json
import pathlib
import os
from typing import Any, Callable

import yaml
from platformdirs import user_data_dir


def token_path(app_name: str, filename: str = "token.yaml") -> str:
    """Get the path to the token file.

    This function returns the path to the token file based on the application name
    and the filename. The token file is stored in the user data directory.

    Args:
        app_name: The application name
        filename: The token file name

    Returns:
        The path to the token file
    """
    data_dir = user_data_dir(app_name)
    pathlib.Path(data_dir).mkdir(parents=True, exist_ok=True)
    return os.path.join(data_dir, filename)


def token_writer(token_path: str) -> Callable[[dict[str, Any], ...], None]:
    """Create a function that writes token data to a file.

    This function creates a token writer that supports both JSON and YAML formats
    based on the file extension. If the filename ends with '.json', JSON format
    will be used; otherwise, YAML format will be used.

    Args:
        token_path: Path to the token file

    Returns:
        A function that takes a token dictionary and writes it to the file
    """

    def write_token(token: dict[str, Any], *args, **kwargs) -> None:
        """Write the token data to a file.

        Args:
            token: The OAuth token data dictionary
            *args: Additional arguments (ignored)
            **kwargs: Additional keyword arguments (ignored)
        """
        if not token:
            return

        with open(token_path, "w") as f:
            if token_path.endswith(".json"):
                return json.dump(token, f)

            # Round Trip the token through JSON to ensure it's serializable
            return yaml.safe_dump(
                json.loads(json.dumps(token)),
                f,
                default_flow_style=False,
                explicit_start=True,
            )

    return write_token


def token_loader(token_path: str) -> Callable[[], dict[str, Any]]:
    """Create a function that loads token data from a file.

    This function creates a token loader that supports both JSON and YAML formats
    based on the file extension. If the filename ends with '.json', JSON format
    will be used; otherwise, YAML format will be used.

    Args:
        token_path: Path to the token file

    Returns:
        A function that loads and returns token data from the file
    """

    def load_token() -> dict[str, Any]:
        """Load the token data from a file.

        Returns:
            The OAuth token data as a dictionary
        """
        with open(token_path, "r") as f:
            if token_path.endswith(".json"):
                return json.load(f)

            return yaml.safe_load(f)

    return load_token


class Manager:
    def __init__(self, path: str):
        self.path = path
        self.load = token_loader(self.path)
        self.write = token_writer(self.path)

    def exists(self) -> bool:
        return os.path.exists(self.path)
