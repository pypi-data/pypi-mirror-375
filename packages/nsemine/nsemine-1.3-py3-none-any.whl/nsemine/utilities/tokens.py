index_tokens = {
    'NIFTY 50': 26000,
    'NIFTY BANK': 26009,
    'NIFTY NEXT 50': 26013,
    'NIFTY FIN SERVICE': 26037,
    'NIFTY FINANCIAL SERVICES': 26037,
    'NIFTY MIDCAP 100': 26010,    
    'NIFTY MID SELECT': 26074,
    'NIFTY MIDCAP SELECT': 26074,
    'NIFTY TOTAL MKT': 26021,
    'NIFTY TOTAL MARKET': 26021,
    'NIFTY IT': 26008,
    'NIFTY 500': 26003,
    'NIFTY 100': 26012,
    'INDIA VIX': 26017,
    'NIFTY MIDCAP 50': 26014,
    'NIFTY REALTY': 26052,
    'NIFTY INFRA': 26019,
    'NIFTY ENERGY': 26054,
    'NIFTY FMCG': 26055,
    'NIFTY MNC': 26056,
    'NIFTY PHARMA': 26057,
    'NIFTY PSE': 26024,
    'NIFTY PSU BANK': 26059,
    'NIFTY SERV SECTOR': 26060,
    'NIFTY SMLCAP 100': 26011,
    'NIFTY SMALLCAP 100': 26011,
    'NIFTY 200': 26065,
    'NIFTY AUTO': 26061,
    'NIFTY MEDIA': 26063,
    'NIFTY METAL': 26062,
    'NIFTY DIV OPPS 50': 26078,
    'NIFTY COMMODITIES': 26066,
    'NIFTY CONSUMPTION': 26048,
    'NIFTY50 DIV POINT': 26080,
    'NIFTY100 LIQ 15': 26051,
    'NIFTY CPSE': 26041,
    'NIFTY GROWSECT 15': 26082,
    'NIFTY50 TR 2X LEV': 26075,
    'NIFTY50 PR 2X LEV': 26064,
    'NIFTY50 TR 1X INV': 26076,
    'NIFTY50 PR 1X INV': 26020,
    'NIFTY50 VALUE 20': 26081,
    'NIFTY100 QUALTY30': 26026,
    'NIFTY MID LIQ 15': 26018,
    'NIFTY PVT BANK': 26025,
    'NIFTY PRIVATE BANK': 26025,
    'NIFTY GS 8 13YR': 26053,
    'NIFTY GS 10YR': 26058,
    'NIFTY GS 10YR CLN': 26029,
    'NIFTY GS 4 8YR': 26030,
    'NIFTY GS 11 15YR': 26069,
    'NIFTY GS 15YRPLUS': 26070,
    'NIFTY GS COMPSITE': 26071,
    'NIFTY50 EQL WGT': 26083,
    'NIFTY100 EQL WGT': 26084,
    'NIFTY100 LOWVOL30': 26085,
    'NIFTY ALPHA 50': 26086,
    'NIFTY MIDCAP 150': 26087,
    'NIFTY SMLCAP 50': 26088,
    'NIFTY SMALLCAP 50': 26088,
    'NIFTY SMLCAP 250': 26089,
    'NIFTY SMALLCAP 250': 26089,
    'NIFTY MIDSML 400': 26091,
    'NIFTY200 QUALTY30': 26092,
    'NIFTY FINSRV25 50': 26093,
    'NIFTY ALPHALOWVOL': 26094,
    'NIFTY200MOMENTM30': 26067,
    'NIFTY100ESGSECLDR': 26068,
    'NIFTY HEALTHCARE': 26096,
    'NIFTY HEALTHCARE INDEX': 26096,
    'NIFTY CONSR DURBL': 26095,
    'NIFTY CONSUMER DURABLES': 26095,
    'NIFTY OIL AND GAS': 26099,
    'NIFTY OIL & GAS' : 26099,
    'NIFTY500 MULTICAP': 26098,
    'NIFTY LARGEMID250': 26097,
    'NIFTY MICROCAP250': 26022,
    'NIFTY MICROCAP 250': 26022,
    'NIFTY IND DIGITAL': 26023,
    'NIFTY100 ESG': 26027,
    'NIFTY M150 QLTY50': 26028,
    'NIFTY INDIA MFG': 26031,
    'NIFTY200 ALPHA 30': 26043,
    'NIFTYM150MOMNTM50': 26042,
    'NIFTY TATA 25 CAP': 26047,
    'NIFTY MIDSML HLTH': 26044,
    'NIFTY MIDSMALL HEALTHCARE' : 26044,
    'NIFTY MULTI MFG': 26045,
    'NIFTY MULTI INFRA': 26046,
    'NIFTY IND DEFENCE': 26038,
    'NIFTY IND TOURISM': 26039,
    'NIFTY CAPITAL MKT': 26040,
    'NIFTY500MOMENTM50': 26049,
    'NIFTYMS400 MQ 100': 26050,
    'NIFTYSML250MQ 100': 26079,
    'NIFTY TOP 10 EW': 26077,
    'NIFTY AQL 30': 26100,
    'NIFTY AQLV 30': 26101,
    'NIFTY EV': 26102,
    'NIFTY HIGHBETA 50': 26103,
    'NIFTY NEW CONSUMP': 26104,
    'NIFTY CORP MAATR': 26105,
    'NIFTY LOW VOL 50': 26106,
    'NIFTY MOBILITY': 26107,
    'NIFTY QLTY LV 30': 26108,
    'NIFTY SML250 Q50': 26109,
    'NIFTY TOP 15 EW': 26110,
    'NIFTY100 ALPHA 30': 26111,
    'NIFTY100 ENH ESG': 26112,
    'NIFTY200 VALUE 30': 26113,
    'NIFTY500 EW': 26114,
    'NIFTY MULTI MQ 50': 26115,
    'NIFTY500 VALUE 50': 26116,
    'NIFTY TOP 20 EW': 26117,
    'NIFTY COREHOUSING': 26118,
    'NIFTY FINSEREXBNK': 26119,
    'NIFTY HOUSING': 26120,
    'NIFTY IPO': 26121,
    'NIFTY MS FIN SERV': 26122,
    'NIFTY MIDSMALL FINANCIAL SERVICES': 26122,
    'NIFTY MS IND CONS': 26123,
    'NIFTY MS IT TELCM': 26124,
    'NIFTY MIDSMALL IT & TELECOM' : 26124,
    'NIFTY NONCYC CONS': 26125,
    'NIFTY RURAL': 26126,
    'NIFTY SHARIAH 25': 26127,
    'NIFTY TRANS LOGIS': 26128,
    'NIFTY50 SHARIAH': 26129,
    'NIFTY500 LMS EQL': 26130,
    'NIFTY500 SHARIAH': 26131,
    'NIFTY500 QLTY50': 26132,
    'NIFTY500 LOWVOL50': 26133,
    'NIFTY500 MQVLV50': 26134,
}