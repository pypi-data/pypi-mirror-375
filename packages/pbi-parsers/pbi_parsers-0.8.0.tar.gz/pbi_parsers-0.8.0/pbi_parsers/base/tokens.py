from dataclasses import dataclass, field
from typing import Any


@dataclass
class TextSlice:
    full_text: str = ""
    start: int = -1
    end: int = -1

    def __eq__(self, other: object) -> bool:
        """Checks equality based on the text slice."""
        if not isinstance(other, TextSlice):
            return NotImplemented
        return self.full_text == other.full_text and self.start == other.start and self.end == other.end

    def __hash__(self) -> int:
        """Returns a hash based on the text slice."""
        return hash((self.full_text, self.start, self.end))

    def __repr__(self) -> str:
        """Returns a string representation of the TextSlice."""
        return f"TextSlice(text='{self.get_text()}', start={self.start}, end={self.end})"

    def get_text(self) -> str:
        """Returns the text slice."""
        return self.full_text[self.start : self.end]


@dataclass
class BaseToken:
    tok_type: Any
    text_slice: TextSlice = field(default_factory=TextSlice)

    def __eq__(self, other: object) -> bool:
        """Checks equality based on token type and text slice."""
        if not isinstance(other, BaseToken):
            return NotImplemented
        return self.tok_type == other.tok_type and self.text_slice == other.text_slice

    def __hash__(self) -> int:
        """Returns a hash based on token type and text slice."""
        return hash((self.tok_type, self.text_slice))

    def __repr__(self) -> str:
        pretty_text = self.text_slice.get_text().replace("\n", "\\n").replace("\r", "\\r")
        return f"Token(type={self.tok_type.name}, text='{pretty_text}')"

    def position(self) -> tuple[int, int]:
        """Returns the start and end positions of the token.

        Returns:
            tuple[int, int]: A tuple containing the start and end positions of the token within the source text.

        """
        return self.text_slice.start, self.text_slice.end

    @property
    def text(self) -> str:
        """Returns the text underlying the token.

        Returns:
            str: The text of the token as a string.

        """
        return self.text_slice.get_text()
