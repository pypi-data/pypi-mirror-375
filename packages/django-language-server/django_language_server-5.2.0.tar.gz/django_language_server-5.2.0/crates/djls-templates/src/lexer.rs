use thiserror::Error;

use crate::tokens::Token;
use crate::tokens::TokenType;

pub struct Lexer {
    source: String,
    chars: Vec<char>,
    start: usize,
    current: usize,
    line: usize,
}

impl Lexer {
    #[must_use]
    pub fn new(source: &str) -> Self {
        Lexer {
            source: String::from(source),
            chars: source.chars().collect(),
            start: 0,
            current: 0,
            line: 1,
        }
    }

    #[allow(clippy::too_many_lines)]
    pub fn tokenize(&mut self) -> Result<Vec<Token>, LexerError> {
        let mut tokens = Vec::new();

        while !self.is_at_end() {
            self.start = self.current;

            let token_type = match self.peek()? {
                '{' => match self.peek_next()? {
                    '%' => {
                        self.consume_n(2)?; // {%
                        let content = self.consume_until("%}")?;
                        self.consume_n(2)?; // %}
                        TokenType::DjangoBlock(content)
                    }
                    '{' => {
                        self.consume_n(2)?; // {{
                        let content = self.consume_until("}}")?;
                        self.consume_n(2)?; // }}
                        TokenType::DjangoVariable(content)
                    }
                    '#' => {
                        self.consume_n(2)?; // {#
                        let content = self.consume_until("#}")?;
                        self.consume_n(2)?; // #}
                        TokenType::Comment(content, "{#".to_string(), Some("#}".to_string()))
                    }
                    _ => {
                        self.consume()?; // {
                        TokenType::Text(String::from("{"))
                    }
                },

                '<' => match self.peek_next()? {
                    '/' => {
                        self.consume_n(2)?; // </
                        let tag = self.consume_until(">")?;
                        self.consume()?; // >
                        TokenType::HtmlTagClose(tag)
                    }
                    '!' if self.matches("<!--") => {
                        self.consume_n(4)?; // <!--
                        let content = self.consume_until("-->")?;
                        self.consume_n(3)?; // -->
                        TokenType::Comment(content, "<!--".to_string(), Some("-->".to_string()))
                    }
                    _ => {
                        self.consume()?; // consume <
                        let tag = self.consume_until(">")?;
                        self.consume()?; // consume >
                        if tag.starts_with("script") {
                            TokenType::ScriptTagOpen(tag)
                        } else if tag.starts_with("style") {
                            TokenType::StyleTagOpen(tag)
                        } else if tag.ends_with('/') {
                            TokenType::HtmlTagVoid(tag.trim_end_matches('/').to_string())
                        } else {
                            TokenType::HtmlTagOpen(tag)
                        }
                    }
                },

                '/' => match self.peek_next()? {
                    '/' => {
                        self.consume_n(2)?; // //
                        let content = self.consume_until("\n")?;
                        TokenType::Comment(content, "//".to_string(), None)
                    }
                    '*' => {
                        self.consume_n(2)?; // /*
                        let content = self.consume_until("*/")?;
                        self.consume_n(2)?; // */
                        TokenType::Comment(content, "/*".to_string(), Some("*/".to_string()))
                    }
                    _ => {
                        self.consume()?;
                        TokenType::Text("/".to_string())
                    }
                },

                c if c.is_whitespace() => {
                    if c == '\n' || c == '\r' {
                        self.consume()?; // \r or \n
                        if c == '\r' && self.peek()? == '\n' {
                            self.consume()?; // \n of \r\n
                        }
                        TokenType::Newline
                    } else {
                        self.consume()?; // Consume the first whitespace
                        while !self.is_at_end() && self.peek()?.is_whitespace() {
                            if self.peek()? == '\n' || self.peek()? == '\r' {
                                break;
                            }
                            self.consume()?;
                        }
                        let whitespace_count = self.current - self.start;
                        TokenType::Whitespace(whitespace_count)
                    }
                }

                _ => {
                    let mut text = String::new();
                    while !self.is_at_end() {
                        let c = self.peek()?;
                        if c == '{' || c == '<' || c == '\n' {
                            break;
                        }
                        text.push(c);
                        self.consume()?;
                    }
                    TokenType::Text(text)
                }
            };

            let token = Token::new(token_type, self.line, Some(self.start));

            match self.peek_previous()? {
                '\n' => self.line += 1,
                '\r' => {
                    self.line += 1;
                    if self.peek()? == '\n' {
                        self.current += 1;
                    }
                }
                _ => {}
            }

            tokens.push(token);
        }

        // Add EOF token
        let eof_token = Token::new(TokenType::Eof, self.line, None);
        tokens.push(eof_token);

        Ok(tokens)
    }

    fn peek(&self) -> Result<char, LexerError> {
        self.peek_at(0)
    }

    fn peek_next(&self) -> Result<char, LexerError> {
        self.peek_at(1)
    }

    fn peek_previous(&self) -> Result<char, LexerError> {
        self.peek_at(-1)
    }

    #[allow(dead_code)]
    fn peek_until(&self, end: &str) -> bool {
        let mut index = self.current;
        let end_chars: Vec<char> = end.chars().collect();

        while index < self.chars.len() {
            if self.chars[index..].starts_with(&end_chars) {
                return true;
            }
            index += 1;
        }
        false
    }

    #[allow(clippy::cast_sign_loss)]
    fn peek_at(&self, offset: isize) -> Result<char, LexerError> {
        // Safely handle negative offsets
        let index = if offset < 0 {
            // Check if we would underflow
            if self.current < offset.unsigned_abs() {
                return Err(LexerError::AtBeginningOfSource);
            }
            self.current - offset.unsigned_abs()
        } else {
            // Safe addition since offset is positive
            self.current + (offset as usize)
        };

        self.item_at(index)
    }

    fn item_at(&self, index: usize) -> Result<char, LexerError> {
        if index >= self.source.len() {
            // Return a null character when past the end, a bit of a departure from
            // idiomatic Rust code, but makes writing the matching above and testing
            // much easier
            Ok('\0')
        } else {
            self.source
                .chars()
                .nth(index)
                .ok_or(LexerError::InvalidCharacterAccess)
        }
    }

    fn matches(&mut self, pattern: &str) -> bool {
        let mut i = self.current;
        for c in pattern.chars() {
            if i >= self.chars.len() || self.chars[i] != c {
                return false;
            }
            i += 1;
        }
        true
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn consume(&mut self) -> Result<char, LexerError> {
        if self.is_at_end() {
            return Err(LexerError::AtEndOfSource);
        }
        self.current += 1;
        self.peek_previous()
    }

    fn consume_n(&mut self, count: usize) -> Result<String, LexerError> {
        let start = self.current;
        for _ in 0..count {
            self.consume()?;
        }
        Ok(self.source[start..self.current].trim().to_string())
    }

    #[allow(dead_code)]
    fn consume_chars(&mut self, s: &str) -> Result<char, LexerError> {
        for c in s.chars() {
            if c != self.peek()? {
                return Err(LexerError::UnexpectedCharacter(c, self.line));
            }
            self.consume()?;
        }
        self.peek_previous()
    }

    fn consume_until(&mut self, s: &str) -> Result<String, LexerError> {
        let start = self.current;
        while !self.is_at_end() {
            if self.chars[self.current..self.chars.len()]
                .starts_with(s.chars().collect::<Vec<_>>().as_slice())
            {
                return Ok(self.source[start..self.current].trim().to_string());
            }
            self.consume()?;
        }
        Err(LexerError::UnexpectedEndOfInput)
    }
}

#[derive(Error, Debug)]
pub enum LexerError {
    #[error("empty token at line {0}")]
    EmptyToken(usize),

    #[error("unexpected character '{0}' at line {1}")]
    UnexpectedCharacter(char, usize),

    #[error("unexpected end of input")]
    UnexpectedEndOfInput,

    #[error("source is empty")]
    EmptySource,

    #[error("at beginning of source")]
    AtBeginningOfSource,

    #[error("at end of source")]
    AtEndOfSource,

    #[error("invalid character access")]
    InvalidCharacterAccess,

    #[error("unexpected token type '{0:?}'")]
    UnexpectedTokenType(TokenType),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize_html() {
        let source = r#"<div class="container" id="main" disabled></div>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_django_variable() {
        let source = "{{ user.name|default:\"Anonymous\"|title }}";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_django_block() {
        let source = "{% if user.is_staff %}Admin{% else %}User{% endif %}";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_comments() {
        let source = r"<!-- HTML comment -->
{# Django comment #}
<script>
    // JS single line comment
    /* JS multi-line
       comment */
</script>
<style>
    /* CSS comment */
</style>";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_script() {
        let source = r#"<script type="text/javascript">
    // Single line comment
    const x = 1;
    /* Multi-line
       comment */
    console.log(x);
</script>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_style() {
        let source = r#"<style type="text/css">
    /* Header styles */
    .header {
        color: blue;
    }
</style>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn test_tokenize_error_cases() {
        // Unterminated tokens
        assert!(Lexer::new("{{ user.name").tokenize().is_err()); // No closing }}
        assert!(Lexer::new("{% if").tokenize().is_err()); // No closing %}
        assert!(Lexer::new("{#").tokenize().is_err()); // No closing #}
        assert!(Lexer::new("<div").tokenize().is_err()); // No closing >

        // Invalid characters or syntax within tokens
        assert!(Lexer::new("{{}}").tokenize().is_ok()); // Empty but valid
        assert!(Lexer::new("{%  %}").tokenize().is_ok()); // Empty but valid
        assert!(Lexer::new("{##}").tokenize().is_ok()); // Empty but valid
    }

    #[test]
    fn test_tokenize_nested_delimiters() {
        let source = r"{{ user.name }}
{% if true %}
{# comment #}
<!-- html comment -->
<div>text</div>";
        assert!(Lexer::new(source).tokenize().is_ok());
    }

    #[test]
    fn test_tokenize_everything() {
        let source = r#"<!DOCTYPE html>
<html>
<head>
    <style type="text/css">
        /* Style header */
        .header { color: blue; }
    </style>
    <script type="text/javascript">
        // Init app
        const app = {
            /* Config */
            debug: true
        };
    </script>
</head>
<body>
    <!-- Header section -->
    <div class="header" id="main" data-value="123" disabled>
        {% if user.is_authenticated %}
            {# Welcome message #}
            <h1>Welcome, {{ user.name|default:"Guest"|title }}!</h1>
            {% if user.is_staff %}
                <span>Admin</span>
            {% else %}
                <span>User</span>
            {% endif %}
        {% endif %}
    </div>
</body>
</html>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize().unwrap();
        insta::assert_yaml_snapshot!(tokens);
    }
}
