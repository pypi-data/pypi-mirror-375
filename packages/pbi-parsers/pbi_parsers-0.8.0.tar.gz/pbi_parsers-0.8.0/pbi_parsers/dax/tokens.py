from dataclasses import dataclass
from enum import Enum, auto

from pbi_parsers.base import BaseToken


class TokenType(Enum):
    AMPERSAND_OPERATOR = auto()
    ASC = auto()
    BRACKETED_IDENTIFIER = auto()
    COMMA = auto()
    COMPARISON_OPERATOR = auto()
    DESC = auto()
    DIVIDE_SIGN = auto()
    DOUBLE_AMPERSAND_OPERATOR = auto()
    DOUBLE_PIPE_OPERATOR = auto()
    EOF = auto()
    EQUAL_SIGN = auto()
    EXPONENTIATION_SIGN = auto()
    FALSE = auto()
    IN = auto()
    LEFT_CURLY_BRACE = auto()
    LEFT_PAREN = auto()
    MINUS_SIGN = auto()
    MODULUS_SIGN = auto()
    MULTIPLY_SIGN = auto()
    MULTI_LINE_COMMENT = auto()
    NOT_EQUAL_SIGN = auto()
    NUMBER_LITERAL = auto()
    PERIOD = auto()
    PLUS_SIGN = auto()
    RETURN = auto()
    RIGHT_CURLY_BRACE = auto()
    RIGHT_PAREN = auto()
    SINGLE_LINE_COMMENT = auto()
    SINGLE_QUOTED_IDENTIFIER = auto()
    STRING_LITERAL = auto()
    TRUE = auto()
    UNQUOTED_IDENTIFIER = auto()
    VARIABLE = auto()
    WHITESPACE = auto()


KEYWORD_MAPPING = {
    "TRUE": TokenType.TRUE,
    "FALSE": TokenType.FALSE,
    "ASC": TokenType.ASC,
    "DESC": TokenType.DESC,
}


@dataclass
class Token(BaseToken):
    tok_type: TokenType = TokenType.EOF
