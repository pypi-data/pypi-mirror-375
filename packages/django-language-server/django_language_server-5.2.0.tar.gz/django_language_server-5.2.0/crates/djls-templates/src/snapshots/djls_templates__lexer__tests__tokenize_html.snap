---
source: crates/djls-templates/src/lexer.rs
expression: tokens
---
- token_type:
    HtmlTagOpen: "div class=\"container\" id=\"main\" disabled"
  line: 1
  start: 0
- token_type:
    HtmlTagClose: div
  line: 1
  start: 42
- token_type: Eof
  line: 1
  start: ~
